beefy-gadget/src/worker.rs

```diff
@@ -55,12 +63,101 @@ where
 	<<B::Header as Header>::Hashing as Hash>::hash(b"beefy")
 }
 
+/// Allows messages from last `MAX_LIVE_GOSSIP_ROUNDS` to flow, everything else gets
+/// rejected/expired. All messaging is handled in a single global topic.
+pub struct BeefyGossipValidator<B, P>
+where
+	B: Block,
+{
+	topic: B::Hash,
+	live_rounds: RwLock<Vec<NumberFor<B>>>,
+	_pair: PhantomData<P>,
+}
+
+impl<B, P> BeefyGossipValidator<B, P>
+where
+	B: Block,
+{
+	pub fn new() -> BeefyGossipValidator<B, P> {
+		BeefyGossipValidator {
+			topic: topic::<B>(),
+			live_rounds: RwLock::new(Vec::new()),
+			_pair: PhantomData,
+		}
+	}
+
+	fn note_round(&self, round: NumberFor<B>) {
+		let mut live_rounds = self.live_rounds.write();
+
+		// NOTE: ideally we'd use a VecDeque here, but currently binary search is only available on
+		// nightly for `VecDeque`.
+		while live_rounds.len() > MAX_LIVE_GOSSIP_ROUNDS {
+			let _ = live_rounds.remove(0);
+		}
+
+		if let Some(idx) = live_rounds.binary_search(&round).err() {
+			live_rounds.insert(idx, round);
```

Comment:
Wouldn't it make sense to use `BTreeSet` instead? Pruning can easily be done with `split_off`.

That said with 5 elements it probably doesn't matter at all performance wise even if we did linear search, but I'd consider it if it makes the code simpler.

Reply:
Currently `BTreeSet` does not provide any methods for poping or looking at the head. I could inspect the contents of the `BTreeSet` with an iterator but then we're back to linear search. What was your idea of using `split_off`? If the set has elements `[1,2,4,5]` and I want to insert `3` I don't actually want to prune anything since it only has 4 elements.

Reply:
I was assuming that eventually the rounds will be consecutive anyway (i.e. we will eventually receive all of them, but maybe in some scrumbled order). So instead of storing up to 5 entries, I would go for storing at most entries from rounds `[N-5; N]`. Hence `split_off` could simply be used to get rid of all old rounds. The inspection would not really be needed if we just stored the best session we've seen so far.

Reply:
> I was assuming that eventually the rounds will be consecutive anyway

Yeah, I think so but I'm not sure it is completely guaranteed with the current logic. The `N` here will be some random block number that fits the power of two rule so we can't easily figure out what the range of values in that vec should be.